### ISSUE TITLE
Exception while downloading content

### ISSUE THREAD

**Issue by eximius313** [playframework/playframework#9178]
### Play Version (2.5.x / etc)
`2.6.21`

### API (Scala / Java / Neither / Both)
Java

### Operating System (Ubuntu 15.10 / MacOS 10.10 / Windows 10)
Windows 7

### JDK (Oracle 1.8.0_72, OpenJDK 1.8.x, Azul Zing)
JDK 9.0.4

### Expected Behavior
No exception is thrown

### Actual Behavior
```
23:06:06.799 [application-akka.actor.default-dispatcher-26] [error] play.core.server.common.ServerResultUtils - Exception occurred while converting Result with headers [<Conte
nt-Disposition>: <attachment; filename*=UTF-8''This is ń.pdf>]. Calling HttpErrorHandler to get alternative Result.
java.lang.IllegalArgumentException: Character 324 cannot match AsciiSet because it is out of range
        at play.core.utils.AsciiBitSet.get(AsciiSet.scala:81)
        at play.core.server.common.ServerResultUtils.loop$1(ServerResultUtils.scala:122)
        at play.core.server.common.ServerResultUtils.validateString(ServerResultUtils.scala:126)
        at play.core.server.common.ServerResultUtils.validateHeaderValueChars(ServerResultUtils.scala:116)
        at play.core.server.akkahttp.AkkaModelConversion.$anonfun$convertHeaders$1(AkkaModelConversion.scala:287)
        at scala.collection.TraversableLike.$anonfun$flatMap$1(TraversableLike.scala:241)
        at scala.collection.immutable.RedBlackTree$._foreach(RedBlackTree.scala:101)
        at scala.collection.immutable.RedBlackTree$.foreach(RedBlackTree.scala:97)
        at scala.collection.immutable.TreeMap.foreach(TreeMap.scala:202)
        at scala.collection.TraversableLike.flatMap(TraversableLike.scala:241)
        at scala.collection.TraversableLike.flatMap$(TraversableLike.scala:238)
        at scala.collection.immutable.TreeMap.flatMap(TreeMap.scala:47)
        at play.core.server.akkahttp.AkkaModelConversion.convertHeaders(AkkaModelConversion.scala:290)
        at play.core.server.akkahttp.AkkaModelConversion.$anonfun$convertResult$2(AkkaModelConversion.scala:200)
        at akka.http.scaladsl.util.FastFuture$.$anonfun$map$1(FastFuture.scala:23)
        at akka.http.scaladsl.util.FastFuture$.strictTransform$1(FastFuture.scala:41)
        at akka.http.scaladsl.util.FastFuture$.transformWith$extension1(FastFuture.scala:55)
        at akka.http.scaladsl.util.FastFuture$.map$extension(FastFuture.scala:23)
        at play.core.server.akkahttp.AkkaModelConversion.$anonfun$convertResult$1(AkkaModelConversion.scala:199)
        at play.core.server.common.ServerResultUtils.resultConversionWithErrorHandling(ServerResultUtils.scala:180)
        at play.core.server.akkahttp.AkkaModelConversion.convertResult(AkkaModelConversion.scala:221)
        at play.core.server.AkkaHttpServer.$anonfun$runAction$5(AkkaHttpServer.scala:386)
        at scala.concurrent.Future.$anonfun$flatMap$1(Future.scala:304)
        at scala.concurrent.impl.Promise.$anonfun$transformWith$1(Promise.scala:37)
        at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:60)
        at akka.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:55)
        at akka.dispatch.BatchingExecutor$BlockableBatch.$anonfun$run$1(BatchingExecutor.scala:91)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
        at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:81)
        at akka.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:91)
        at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:40)
        at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:44)
        at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
        at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
        at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
        at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
23:06:06.822 [application-akka.actor.default-dispatcher-26] [error] application -

! @7beoi050g - Internal server error, for (GET) [/download] ->

play.api.UnexpectedException: Unexpected exception[ServerResultException: Error converting Play Result for server backend]
        at play.api.http.HttpErrorHandlerExceptions$.throwableToUsefulException(HttpErrorHandler.scala:247)
        at play.api.http.DefaultHttpErrorHandler.onServerError(HttpErrorHandler.scala:178)
        at play.core.server.common.ServerResultUtils.play$core$server$common$ServerResultUtils$$handleConversionError$1(ServerResultUtils.scala:161)
        at play.core.server.common.ServerResultUtils$$anonfun$resultConversionWithErrorHandling$9.applyOrElse(ServerResultUtils.scala:180)
        at play.core.server.common.ServerResultUtils$$anonfun$resultConversionWithErrorHandling$9.applyOrElse(ServerResultUtils.scala:180)
        at scala.concurrent.Future.$anonfun$recoverWith$1(Future.scala:414)
        at akka.http.scaladsl.util.FastFuture$.$anonfun$transformWith$2(FastFuture.scala:37)
        at akka.http.scaladsl.util.FastFuture$.strictTransform$1(FastFuture.scala:41)
        at akka.http.scaladsl.util.FastFuture$.transformWith$extension1(FastFuture.scala:46)
        at akka.http.scaladsl.util.FastFuture$.transformWith$extension0(FastFuture.scala:37)
        at akka.http.scaladsl.util.FastFuture$ErrorFuture.transformWith(FastFuture.scala:97)
        at scala.concurrent.Future.recoverWith(Future.scala:413)
        at scala.concurrent.Future.recoverWith$(Future.scala:412)
        at akka.http.scaladsl.util.FastFuture$ErrorFuture.recoverWith(FastFuture.scala:88)
        at play.core.server.common.ServerResultUtils.resultConversionWithErrorHandling(ServerResultUtils.scala:180)
        at play.core.server.akkahttp.AkkaModelConversion.convertResult(AkkaModelConversion.scala:221)
        at play.core.server.AkkaHttpServer.$anonfun$runAction$5(AkkaHttpServer.scala:386)
        at scala.concurrent.Future.$anonfun$flatMap$1(Future.scala:304)
        at scala.concurrent.impl.Promise.$anonfun$transformWith$1(Promise.scala:37)
        at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:60)
        at akka.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:55)
        at akka.dispatch.BatchingExecutor$BlockableBatch.$anonfun$run$1(BatchingExecutor.scala:91)
        at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
        at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:81)
        at akka.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:91)
        at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:40)
        at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:44)
        at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
        at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
        at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
        at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
Caused by: play.core.server.common.ServerResultException: Error converting Play Result for server backend
        ... 29 common frames omitted
Caused by: java.lang.IllegalArgumentException: Character 324 cannot match AsciiSet because it is out of range
        at play.core.utils.AsciiBitSet.get(AsciiSet.scala:81)
        at play.core.server.common.ServerResultUtils.loop$1(ServerResultUtils.scala:122)
        at play.core.server.common.ServerResultUtils.validateString(ServerResultUtils.scala:126)
        at play.core.server.common.ServerResultUtils.validateHeaderValueChars(ServerResultUtils.scala:116)
        at play.core.server.akkahttp.AkkaModelConversion.$anonfun$convertHeaders$1(AkkaModelConversion.scala:287)
        at scala.collection.TraversableLike.$anonfun$flatMap$1(TraversableLike.scala:241)
        at scala.collection.immutable.RedBlackTree$._foreach(RedBlackTree.scala:101)
        at scala.collection.immutable.RedBlackTree$.foreach(RedBlackTree.scala:97)
        at scala.collection.immutable.TreeMap.foreach(TreeMap.scala:202)
        at scala.collection.TraversableLike.flatMap(TraversableLike.scala:241)
        at scala.collection.TraversableLike.flatMap$(TraversableLike.scala:238)
        at scala.collection.immutable.TreeMap.flatMap(TreeMap.scala:47)
        at play.core.server.akkahttp.AkkaModelConversion.convertHeaders(AkkaModelConversion.scala:290)
        at play.core.server.akkahttp.AkkaModelConversion.$anonfun$convertResult$2(AkkaModelConversion.scala:200)
        at akka.http.scaladsl.util.FastFuture$.$anonfun$map$1(FastFuture.scala:23)
        at akka.http.scaladsl.util.FastFuture$.strictTransform$1(FastFuture.scala:41)
        at akka.http.scaladsl.util.FastFuture$.transformWith$extension1(FastFuture.scala:55)
        at akka.http.scaladsl.util.FastFuture$.map$extension(FastFuture.scala:23)
        at play.core.server.akkahttp.AkkaModelConversion.$anonfun$convertResult$1(AkkaModelConversion.scala:199)
        ... 17 common frames omitted
```

### Reproducible Test Case
```
byte[] data = {};//pdf
String name = "This is ń";
return ok(data).as("application/pdf")
                .withHeader("Content-Disposition", "attachment; filename*=UTF-8''"+name)
```

**Comment by mkurz** [comment]
@eximius313 In case you have not solved this yet, the solution is to encode the filename:
```java
String name = java.net.URLEncoder.encode("This is ń", "utf-8");
```
(utf-8 is used because that is what you use in the `Content-Disposition` header as well.

This is not a bug in Play, this is just how it works. You must encode filenames in such cases.

**Comment by eximius313** [comment]
But then the result is: `This+is+ń` which is not exactly what I want...
Do I really need to do `URLEncoder.encode("This is ń", "UTF-8").replaceAll("+", StringUtils.SPACE);` manually ?

**Comment by mkurz** [comment]
@eximius313 Sorry, I was too fast... Here is the solution now:

For Play 2.7:

```java
import play.core.utils.HttpHeaderParameterEncoding;
import play.mvc.Http.HeaderNames;
```
```java
byte[] data = {};//pdf
String name = HttpHeaderParameterEncoding.encode("filename", "This is ń.pdf");
return ok(data).as("application/pdf")
               .withHeader(HeaderNames.CONTENT_DISPOSITION, "attachment; " + filename);
```

In Play 2.8, thanks to #9589, you will be able to just write:

```java
return ok().sendBytes(data, false, Optional.of("This is ń.pdf"));
```
Also in Play 2.8 I added a helper to create the `Content-Disposition` header: #9574

**Comment by eximius313** [comment]
Thanks @mkurz !

**Comment by mkurz** [comment]
You are welcome @eximius313!

**Comment by sarthakm21** [comment]
@mkurz what is the scala solution for this problem for v2.7?

**Comment by sarthakm21** [comment]
Thank you for your help
I also found this Results.contentDispositionHeader method, added in [this commit](https://github.com/playframework/playframework/commit/2a555d3eb505bf69021d569a9043e7f5f855c194#diff-2b6e4f0b37ca44c5cf997b553a4873f76a9bfc8acd5ccc831c5e53a2ba320453)
Using this should accomplish this task too, right?

**Comment by mkurz** [comment]
Yeah I think that method should do the job as well. But not sure if it is available in the latest 2.7.x release.
